File: 2022-04/python/494_Target_Sum.py

class Solution:

    def findTargetSumWays(self, nums: List[int], target: int) -> int:
        # DFS with cache
        cache = {}
        def dfs(i: int, s: int) -> int:
            cache_key = f"{i}_{s}"
            if cache_key in cache:
                return cache.get(cache_key)
            if i == len(nums):
                return 1 if s == target else 0
            cache[cache_key] = dfs(i + 1, s + nums[i]) + dfs(i + 1, s - nums[i])
            return cache.get(cache_key)
        return dfs(0, 0)
    
    # TODO with DP

File: 2022-04/typescript/494_Target_Sum.ts

function findTargetSumWays(nums: number[], target: number): number {
    const cache = new Map<string, number>()
    function dfs(i: number, s: number): number {
        const cacheKey = [i, s].join('_')
        if (cache.has(cacheKey)) {
            return cache.get(cacheKey)
        }
        if (i === nums.length) {
            return s === target ? 1 : 0
        }
        cache.set(cacheKey, dfs(i + 1, s + nums[i]) + dfs(i + 1, s - nums[i]))
        return cache.get(cacheKey)
    }
    return dfs(0, 0)
}

// TODO with DP